object.c

```diff
@@ -4080,6 +4083,25 @@ f_sprintf(int c, const VALUE *v, VALUE _)
     return rb_f_sprintf(c, v);
 }
 
+static VALUE
+rb_immutable_new(int argc, VALUE *argv, VALUE self)
+{
+    VALUE result = rb_call_super(argc, argv);
+
+    rb_funcall(result, idFreeze, 0);
+    // rb_obj_freeze(result);
```

Comment:
We should do an internal deep freeze here, see https://bugs.ruby-lang.org/issues/18035?next_issue_id=18034&prev_issue_id=18036#note-21

Reply:
Lots of use cases for `freeze` will generate caches and other optimised data structures BEFORE being frozen so some how we need to handle this.

Reply:
Yes, so call user `freeze` + deep freeze internally after to ensure everything is actually frozen.

Reply:
Okay, I'll try this.